### PR TITLE
[Models] Add tentacle and package last version

### PR DESCRIPTION
**Pull request by Herklos**

### CODE REVIEW

File: app/util/spree/tentacle_version_comparator.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module Spree
+  class TentacleVersionComparator < Gem::Version
+    LATEST_VERSION = "latest"
+
+    def self.correct?(version)
+      !!(version.to_s =~ ANCHORED_VERSION_PATTERN or version.to_s == LATEST_VERSION)
+    end
+
+    def <=>(other)
```
Comment (GuillaumeDSM) [inline app/util/spree/tentacle_version_comparator.rb:9]: 👍

File: db/migrate/20201130207807_create_spree_tentacle_package_versions.rb [hunk 2]
```diff
@@ -7,6 +7,5 @@ def change
 
       t.timestamps
     end
-    add_index :updated_at # improve 'last version' query
```
Comment (GuillaumeDSM) [inline db/migrate/20201130207807_create_spree_tentacle_package_versions.rb:10]: 👍

File: spec/factories/tentacles_registry_factory.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :tentacles_registry, class: Spree::TentaclesRegistry do
+    registry_type  { "nexus" }
+    name  { generate(:random_string) }
+    url  { Faker::Internet.url }
+  end
+end
```
Comment (GuillaumeDSM) [inline spec/factories/tentacles_registry_factory.rb:9]: 👍

File: spec/models/spree/tentacle_package_spec.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+require 'spec_helper'
+
+describe Spree::TentaclePackage, type: :model do
+  let(:tentacles_registry) { FactoryBot.create(:tentacles_registry) }
+  let(:tentacle_package) { FactoryBot.create(:tentacle_package, tentacles_registry: tentacles_registry) }
+  let(:tentacle_package_version) { FactoryBot.create(:tentacle_package_version, tentacle_package: tentacle_package) }
+
+  describe '#after_create related tentacle_package_version' do
+    it "when adding first package version" do
```
Comment (GuillaumeDSM) [inline spec/models/spree/tentacle_package_spec.rb:9]: should we add different cases to ensure the "unless" statement is correct ?

File: spec/support/capybara.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    Capybara::Selenium::Driver.new app,
+      browser: :chrome,
+      options: Selenium::WebDriver::Chrome::Options.new(args: %w[headless disable-gpu window-size=1920,1080])
+  end
+  Capybara.javascript_driver = :chrome
+
+  Capybara::Screenshot.register_driver(:chrome) do |driver, path|
+    driver.browser.save_screenshot(path)
+  end
+end
```
Comment (GuillaumeDSM) [inline spec/support/capybara.rb:17]: we need this ?

File: spec/util/spree/tentacle_version_comparator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+require 'spec_helper'
+
+describe Spree::TentacleVersionComparator do
+  LARGE_VERSION = Faker::App.semantic_version(major: 8)
+  MEDIUM_VERSION = Faker::App.semantic_version(major: 2)
+  SMALL_VERSION = Faker::App.semantic_version(major: 0)
+  LATEST_VERSION = "latest"
+  INCORRECT_VERSION = Faker::String.random(length: 6)
+
+  describe '#new compare version' do
```
Comment (GuillaumeDSM) [inline spec/util/spree/tentacle_version_comparator_spec.rb:10]: 💯

File: spec/models/spree/tentacle_package_spec.rb (hunk 4 above)
Comment (Herklos) [reply spec/models/spree/tentacle_package_spec.rb:9]: Yes

File: spec/support/capybara.rb (hunk 5 above)
Comment (Herklos) [reply spec/support/capybara.rb:17]: No, I will remove this :+1:

File: app/models/spree/tentacle_package_version.rb [hunk 7]
```diff
@@ -1,9 +1,19 @@
...
 
   has_many :tentacle_package_version_items, dependent: :destroy, :class_name => 'Spree::TentaclePackageVersionItem'
   has_many :tentacle_versions, :class_name => 'Spree::TentacleVersion', through: :tentacle_package_version_items
   has_many :tentacles, :class_name => 'Spree::Tentacle', through: :tentacle_versions
 
   validates :version, presence: true
+
+  private
+  def update_package_last_version
+    unless self.tentacle_package.last_tentacle_package_version.nil? or Spree::TentacleVersionComparator.new(self.tentacle_package.last_tentacle_package_version.version) > Spree::TentacleVersionComparator.new(self.version)
```
[the remark this answers is not in the document]
Comment (Herklos) [reply app/models/spree/tentacle_package_version.rb:14]: That's better :+1:  I'll change it

File: app/models/spree/tentacle_version.rb [hunk 8]
```diff
@@ -1,10 +1,24 @@
...
   validates :tentacle_id, presence: true
   validates :version, presence: true
   validates :author, presence: true
+
+  private
+
+  def update_tentacle_last_version
+    unless tentacle.last_tentacle_version.nil? \
+      || (Spree::TentacleVersionComparator.new(tentacle.last_tentacle_version.version) \
+      > Spree::TentacleVersionComparator.new(version))
```
Comment (GuillaumeDSM) [inline app/models/spree/tentacle_version.rb:19]: superior here and inferior in TentaclePackageVersion ?

File: spec/models/spree/tentacle_package_spec.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+      it "doesn't update the last_package_version" do
+        FactoryBot.create(:tentacle_package_version,
+                          tentacle_package: tentacle_package,
+                          version: Faker::App.semantic_version(major: 5))
+        expect(tentacle_package.last_tentacle_package_version).to be latest_package_version
+      end
+    end
+  end
+end
```
Comment (GuillaumeDSM) [inline spec/models/spree/tentacle_package_spec.rb:58]: 👍

File: app/models/spree/tentacle_version.rb (hunk 8 above)
Comment (Herklos) [reply app/models/spree/tentacle_version.rb:19]: Yes I forgot to migrate unless to if as in TentaclePackageVersion.

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Install packages
+        run: |
+          yarn install --pure-lockfile
+
+      - name: Run linters
+        run: |
+          bundle exec rubocop --parallel
```
Comment (GuillaumeDSM) [inline .github/workflows/main.yml:53]: 👍

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+#          stylelint TODO uncomment when adding JS
+#          prettier
+#          eslint
+      - name: Run security checks
+        run: |
+          bundle exec brakeman -q -w2
```
Comment (GuillaumeDSM) [inline .github/workflows/main.yml:59]: 👍

File: spec/models/spree/tentacle_version_spec.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+      it "doesn't update the last_tentacle_version" do
+        FactoryBot.create(:tentacle_version,
+                          tentacle: tentacle,
+                          version: Faker::App.semantic_version(major: 5))
+        expect(tentacle.last_tentacle_version).to eq(latest_tentacle_version)
+      end
+    end
+  end
+end
```
Comment (GuillaumeDSM) [inline spec/models/spree/tentacle_version_spec.rb:57]: 👍

File: spec/support/database_cleaner.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+  config.before do
+    DatabaseCleaner.start
+  end
+
+  # After each spec clean the database.
+  config.after do
+    DatabaseCleaner.clean
+  end
+end
```
Comment (GuillaumeDSM) [inline spec/support/database_cleaner.rb:25]: 👍

**Review by GuillaumeDSM — APPROVED** [review]
👍

**Pull request merged**